### PR TITLE
monitoring: T3872: Add a new feature service monitoring

### PR DESCRIPTION
**Pull request by sever-sever**
## Related PR
https://github.com/vyos/vyos-build/pull/192

## Change Summary

Add configurable monitoring service (telegraf parameters)
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T3872

## Component(s) name

telegraf, monitoring
## How to test
VyOS configuration:
```
set service monitoring telegraf authentication organization 'log@in.local'
set service monitoring telegraf authentication token 'GuRJc12tIzfjnYdKRAIYbxdWd2aTpOT9PVYNddzDnFV4HkAcD7u7-kndTFXjGuXzJN6TTxmrvPODB4mnFcseDV=='
set service monitoring telegraf port '8086'
set service monitoring telegraf url 'https://foo.local'

```
It generates config for telegraf
```
vyos@r1-roll# sudo cat /run/telegraf/vyos-telegraf.conf 
# Generated by /usr/libexec/vyos/conf_mode/service_monitoring_telegraf.py

[agent]
  interval = "10s"
  round_interval = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  collection_jitter = "0s"
  flush_interval = "10s"
  flush_jitter = "0s"
  precision = ""
  debug = false
  quiet = false
  logfile = ""
  hostname = ""
  omit_hostname = false
[[outputs.influxdb_v2]]
  urls = ["https://foo.local:8086"]
  insecure_skip_verify = true
  token = "GuRJc12tIzfjnYdKRAIYbxdWd2aTpOT9PVYNddzDnFV4HkAcD7u7-kndTFXjGuXzJN6TTxmrvPODB4mnFcseDV=="
  organization = "log@in.local"
  bucket = "main"
[[inputs.cpu]]
    percpu = true
    totalcpu = true
    collect_cpu_time = false
    report_active = false
[[inputs.disk]]
    ignore_fs = ["devtmpfs", "devfs"]
[[inputs.diskio]]
[[inputs.mem]]
[[inputs.net]]
[[inputs.system]]
[[inputs.netstat]]
[[inputs.processes]]
[[inputs.kernel]]
[[inputs.interrupts]]
[[inputs.linux_sysctl_fs]]
[[inputs.systemd_units]]
[[inputs.conntrack]]
  files = ["ip_conntrack_count","ip_conntrack_max","nf_conntrack_count","nf_conntrack_max"]
  dirs = ["/proc/sys/net/ipv4/netfilter","/proc/sys/net/netfilter"]
[[inputs.ethtool]]
[[inputs.iptables]]
  use_sudo = false
  table = "filter"
  chains = ['VYATTA_PRE_FW_IN_HOOK', 'VYATTA_PRE_FW_FWD_HOOK', 'VYATTA_PRE_FW_OUT_HOOK', 'INPUT', 'FORWARD', 'OUTPUT', 'VYATTA_POST_FW_IN_HOOK', 'VYATTA_POST_FW_FWD_HOOK', 'VYATTA_POST_FW_OUT_HOOK']
  use_lock = true
[[inputs.ntpq]]
  dns_lookup = true
[[inputs.internal]]
[[inputs.nstat]]
[[inputs.syslog]]
  server = "unixgram:///run/telegraf/telegraf_syslog.sock"
  best_effort = true
  syslog_standard = "RFC3164"
[[inputs.exec]]
  commands = [
    "/etc/telegraf/custom_scripts/show_interfaces_input_filter.py",
    "cat /tmp/vyos_services_input_filter"
  ]
  timeout = "10s"
  data_format = "influx"
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by c-po** [comment]
Hi @sever-sever,

a comment on the following CLI commands:

* `set service monitoring telegraf url 'foo.local:8086'`
Please do not place two discrete types in one variable - this will bite back heavily on the validators. Just use a dedicated URL CLI node and then include the generic port XML definition.

* `set service monitoring telegraf authentication login|token`
When I use a token - do I also need to use a login? Or can I use login without token, or token without login?

Please also add the missing `telegraf` dependency to the `debian/control` file

I do not use telegraph, but I bet some users will like it. Thanks!

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                </children>
+              </node>
+              <leafNode name="bucket">
+                <properties>
+                  <help>Remote bucket, by default equal to a hostname</help>
+                </properties>
+              </leafNode>
+              <leafNode name="url">
+                <properties>
+                  <help>Remote url</help>
```
Comment (c-po) [inline interface-definitions/service_monitoring_telegraf.xml.in]: Maybe design a URL validator and use it here.

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                </properties>
+                <children>
+                  <leafNode name="login">
+                    <properties>
+                      <help>Login</help>
+                    </properties>
+                  </leafNode>
+                  <leafNode name="token">
+                    <properties>
+                      <help>Token</help>
```
Comment (c-po) [inline interface-definitions/service_monitoring_telegraf.xml.in]: please add:

```
<valueHelp>
  <format>txt</format>
  <description>A token is a foo bar baz</description>
</valueHelp>
```

Is there a restriction in token characters - like alphabetical only? Please check for a regex pattern.

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            </properties>
+            <children>
+              <node name="authentication">
+                <properties>
+                  <help>Authentication parameters</help>
+                </properties>
+                <children>
+                  <leafNode name="login">
+                    <properties>
+                      <help>Login</help>
```
Comment (c-po) [inline interface-definitions/service_monitoring_telegraf.xml.in]: Please add a:

```
<valueHelp>
  <format>txt</format>
  <description>Login is used for foo bar baz</description>
</valueHelp>
```

Is there a restriction in login names? Please check for a regex pattern.

File: src/conf_mode/service_monitoring_telegraf.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        monitoring['custom_scripts_dir'] = custom_scripts_dir
+
+    return monitoring
+
+def verify(monitoring):
+    # bail out early - looks like removal from running config
+    if not monitoring:
+        return None
+
+    if 'login' not in monitoring['authentication'] or 'token' not in monitoring['authentication']:
```
Comment (c-po) [inline src/conf_mode/service_monitoring_telegraf.py]: You will hit a `KeyError` if `authentication` is not defined.

`if 'authentication' not in monitoring or 'login' not in monitoring['authentication'] or 'token' not in monitoring['authentication']:`

File: src/conf_mode/service_monitoring_telegraf.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+    if 'url' not in monitoring:
+        raise ConfigError(f'Monitoring "url" is mandatory!')
+
+    return None
+
+def generate(monitoring):
+    if not monitoring:
+        # Delete config file
+        if os.path.isfile(config_telegraf):
```
Comment (c-po) [inline src/conf_mode/service_monitoring_telegraf.py]: Maybe build a list of configs and then iterate`

```python
config_files = [config_telegraf, systemd_override, syslog_telegraf]
for file in config_files:
    if os.path.isfile(file):
        os.unlink(file)
```

File: src/conf_mode/service_monitoring_telegraf.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+def apply(monitoring):
+    # Reload systemd manager configuration
+    call('systemctl daemon-reload')
+    if monitoring:
+        call('systemctl restart vyos-telegraf.service')
+    else:
+        call('systemctl stop vyos-telegraf.service')
+    # Telegraf include custom rsyslog config changes
+    call('systemctl restart rsyslog')
```
Comment (c-po) [inline src/conf_mode/service_monitoring_telegraf.py:144]: Is there probably also a reload target for syslog to re-read the config only?

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: An example of URL ```https://foo.local:8086```

**Comment by sever-sever** [comment]
>     * When I use a token - do I also need to use a login? Or can I use login without token, or token without login?

For influxdb2 you should to use "organization" and "token".
I'll update PR to change **login** to **organization**

>    * Please also add the missing telegraf dependency to the debian/control file

The pkg **telegraf** is already in 1.3/1.4:
```
vyos@r1-roll# run show version all | match tele
ii  telegraf                         1.16.2-1                       amd64        Plugin-driven server agent for reporting metrics into InfluxDB.

vyos@r4-epa1:~$ show version all | match teleg
ii  telegraf                             1.16.2-1                            amd64        Plugin-driven server agent for reporting metrics into InfluxDB.
```

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 2 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Additional regex is added.

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 3 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Login was replaced with "organization"
Done

File: src/conf_mode/service_monitoring_telegraf.py (hunk 4 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py]: Done

File: src/conf_mode/service_monitoring_telegraf.py (hunk 5 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py]: Done

**Comment by sever-sever** [comment]
>     * When I use a token - do I also need to use a login? Or can I use login without token, or token without login?

For influxdb2 you should to use "organization" and "token".
I'll update PR to change **login** to **organization**

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 1 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Added regex for URL

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                    <properties>
+                      <help>Authentication token for InfluxDB v2 [REQUIRED]</help>
+                      <valueHelp>
+                        <format>txt</format>
+                        <description>Authentication token</description>
+                      </valueHelp>
+                      <constraint>
+                        <regex>^[a-zA-Z0-9-_]{86}==$</regex>
+                      </constraint>
+                      <constraintErrorMessage>Token contains illegal characters or\nless then 8 characters or\nexceeds 100 character limitation.</constraintErrorMessage>
```
Comment (dmbaturin) [inline interface-definitions/service_monitoring_telegraf.xml.in]: Maybe reword to something like "Token must be 8-100 characters long and must contain only [a-zA-Z0-9-_] characters" to make the error more "actionable"?

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    <children>
+      <node name="monitoring">
+        <properties>
+          <help>Monitoring services</help>
+          <priority>1280</priority>
+        </properties>
+        <children>
+          <node name="telegraf" owner="${vyos_conf_scripts_dir}/service_monitoring_telegraf.py">
+            <properties>
+              <help>Service telegraf</help>
```
Comment (dmbaturin) [inline interface-definitions/service_monitoring_telegraf.xml.in]: Maybe "Telegraf monitoring"?

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    <format>network</format>
+                    <description>Network parameters (net, netstat, nftables)</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>system</format>
+                    <description>System parameters (system, processes, interrupts)</description>
+                  </valueHelp>
+                  <valueHelp>
+                    <format>telegraf</format>
+                    <description>Telegraf parameters</description>
```
Comment (dmbaturin) [inline interface-definitions/service_monitoring_telegraf.xml.in]: Maybe elaborate on it a bit. What are telegraf parameters?

File: interface-definitions/service_monitoring_telegraf.xml.in [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                  <constraint>
+                    <regex>^(all|hardware-utilization|logs|network|system|telegraf)$</regex>
+                  </constraint>
+                  <multi/>
+                </properties>
+                <defaultValue>all</defaultValue>
+              </leafNode>
+              <leafNode name="url">
+                <properties>
+                  <help>Remote url [REQUIRED]</help>
```
Comment (dmbaturin) [inline interface-definitions/service_monitoring_telegraf.xml.in]: "URL" should be uppercase everywhere since it's an abbreviation.

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import unittest
+
+from base_vyostest_shim import VyOSUnitTestSHIM
+
+from vyos.configsession import ConfigSession
+from vyos.configsession import ConfigSessionError
+#from vyos.util import cmd
```
Comment (dmbaturin) [inline smoketest/scripts/cli/test_service_monitoring_telegraf.py]: Nitpick: should we have commented-out imports?

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+from vyos.util import process_named_running
+from vyos.util import read_file
+#from vyos.template import inc_ip
+
+PROCESS_NAME = 'telegraf'
+TELEGRAF_CONF = '/run/telegraf/vyos-telegraf.conf'
+base_path = ['service', 'monitoring', 'telegraf']
+org = 'log@in.local'
+token = 'GuRJc12tIzfjnYdKRAIYbxdWd2aTpOT9PVYNddzDnFV4HkAcD7u7-kndTFXjGuXzJN6TTxmrvPODB4mnFcseDV=='
```
Comment (dmbaturin) [inline smoketest/scripts/cli/test_service_monitoring_telegraf.py:30]: What is this hardcoded token and where it comes from?

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+port = '8888'
+url = 'https://foo.local'
```
Comment (dmbaturin) [inline smoketest/scripts/cli/test_service_monitoring_telegraf.py:32]: Can we avoid fake default URLs?

File: src/conf_mode/service_monitoring_telegraf.py [hunk 14]
```diff
@@ -89,20 +97,21 @@ def verify(monitoring):
...
+            rmtree(cache_dir, ignore_errors=True)
 
         return None
 
     # Create telegraf cache dir
     if not os.path.exists(cache_dir):
-        os.mkdir(cache_dir)
+        os.makedirs(cache_dir)
 
     chown(cache_dir, 'telegraf', 'vyattacfg')
```
Comment (dmbaturin) [inline src/conf_mode/service_monitoring_telegraf.py]: While this is unrelated to the issue on hand, I feel uneasy about hardcoding `vyattacfg` in new scripts. Can we make it a constant somewhere in `vyos.util`, or a function like `make_vyos_group_writable`?

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        index += 1
+        while len(interfaces[index].split())==1:
+            ip = interfaces[index].split()
+            ip_list = ip_list + " " + ip[0]
+            index += 1
+            if index == len(interfaces):
+                break
+    return ip_list
+
+interfaces = subprocess.check_output("/usr/libexec/vyos/op_mode/show_interfaces.py --action=show-brief", shell=True).decode('utf-8').splitlines()
```
Comment (dmbaturin) [inline src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py:34]: For this we do have an environment variable, `vyos_op_scripts_dir`. Or telegraf runs scripts in a way that cannot read the normal shell environment?

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+del interfaces[:3]
+lines_count=len(interfaces)
+index=0
+while index<lines_count:
+    line=interfaces[index].split()
+    if len(line)>1:
+        print("show_interfaces,interface="+line[0]+" ip_addresses=\""+gen_ip_list(index,interfaces)+"\",state="+status_to_int(line[2][0])+"i,link="+status_to_int(line[2][2])+"i,description=\""+description_check(line)+"\" "+str(int(time.time()))+"000000000")
```
Comment (dmbaturin) [inline src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py]: Can we use `"...".format()` here instead of concatenation? It would be easier to both read and edit.

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py (hunk 15 above)
Comment (dmbaturin) [reply src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py:34]: Also, does it send that output unchanged? That script has a header. Shouldn't we make a machine-readable output option for it? I'm happy to assist.

```
Codes: S - State, L - Link, u - Up, D - Down, A - Admin Down
Interface        IP Address                        S/L  Description
---------        ----------                        ---  -----------
```

File: src/etc/telegraf/custom_scripts/vyos_services_input_filter.py [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+# Get running process object
+active_pocess_obj = process_iter()
+for p in active_pocess_obj:
+    for conf_service in conf_services:
+        if conf_service in p.cmdline():
+            # print(conf_service)
+            conf_services[conf_service] = 1
+for conf_service in conf_services:
+    # Printing data in Influx Line Protocol format
+    print("vyos_services,service=\""+conf_service+"\" status="+str(conf_services[conf_service])+"i "+str(int(time.time()))+"000000000")
```
Comment (dmbaturin) [inline src/etc/telegraf/custom_scripts/vyos_services_input_filter.py]: This should also use a format string.

Also, what is `str(int(time.time()))+"000000000")` supposed to do? If it's to signify specific precision, this can also be done with integer formatting in format strings.

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 7 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Sure, it was an old different regex before. 
I'll replace the "constraint message".

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py (hunk 12 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_monitoring_telegraf.py:30]: It was generated in a test environment and part of symbols was replaced
It is just for checks for "smoketest"

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py (hunk 13 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_monitoring_telegraf.py:32]: It is just for "smoketest", org/token/URL parameters necessary, as we compare such parameters with generated config.
So it is not used anywhere in real configs.

File: src/conf_mode/service_monitoring_telegraf.py (hunk 6 above)
Comment (andriiandrieiev) [reply src/conf_mode/service_monitoring_telegraf.py:144]: If you are asking about rsyslog the answer is "no", unfortunately. Meanwhile telegraf can re-read own config easily with "systemctl reload".

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 9 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Telegraf internal statistics
https://github.com/influxdata/telegraf/tree/master/plugins/inputs/internal

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 10 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Done

File: smoketest/scripts/cli/test_service_monitoring_telegraf.py (hunk 11 above)
Comment (sever-sever) [reply smoketest/scripts/cli/test_service_monitoring_telegraf.py]: Fixed

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 7 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Done

File: interface-definitions/service_monitoring_telegraf.xml.in (hunk 8 above)
Comment (sever-sever) [reply interface-definitions/service_monitoring_telegraf.xml.in]: Done

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py (hunk 16 above)
Comment (andriiandrieiev) [reply src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py]: Yes, it will be better.

File: src/etc/telegraf/custom_scripts/vyos_services_input_filter.py (hunk 17 above)
Comment (andriiandrieiev) [reply src/etc/telegraf/custom_scripts/vyos_services_input_filter.py]: It was temporary solution to satisfy datetime format (unix nanoseconds). To solve this we should decide either change Telegraf service precision from nanoseconds to seconds or apply integer formatting.
Personally, I tend to the second option.

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py (hunk 15 above)
Comment (andriiandrieiev) [reply src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py:34]: Yes, output ain't modifying. Actually it's a little bit bigger question. We have two custom scripts which run by Telegraf exec input filter: one gets VyOS NICs list, another tracks VyOS-specific services statuses which we can not track with systemd.  
Probably it would be better to have such scripts inside VyOS, which would provide output in InfluxDB line protocol.

File: src/conf_mode/service_monitoring_telegraf.py (hunk 14 above)
Comment (sever-sever) [reply src/conf_mode/service_monitoring_telegraf.py]: **vyattacfg** was replaced to group **telegraf**

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py (hunk 15 above)
Comment (sever-sever) [reply src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py:34]: @dmbaturin should I change dir for telegraf `custom_scripts` to `/usr/libexec/vyos/op_mode`? It is not related to CLI op-mode commands.

**Comment by lyricconch** [comment]
hi, is there any progress on this topic?

File: src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py (hunk 16 above)
Comment (sever-sever) [reply src/etc/telegraf/custom_scripts/show_interfaces_input_filter.py]: > Can we use `"...".format()` here instead of concatenation? It would be easier to both read and edit.

Replaced with `fstring`

File: src/etc/telegraf/custom_scripts/vyos_services_input_filter.py (hunk 17 above)
Comment (sever-sever) [reply src/etc/telegraf/custom_scripts/vyos_services_input_filter.py]: Replaced to `fstring` format
